processor/intervalprocessor/processor.go

```diff
@@ -52,5 +90,209 @@ func (p *Processor) Capabilities() consumer.Capabilities {
+					return true
+				case pmetric.MetricTypeExponentialHistogram:
+					expHistogram := m.ExponentialHistogram()
+
+					if expHistogram.AggregationTemporality() != pmetric.AggregationTemporalityCumulative {
+						return false
+					}
+
+					aggregateDataPoints(expHistogram.DataPoints(), p.expHistograms, rm.Resource(), rm.SchemaUrl(), sm.Scope(), sm.SchemaUrl(), m)
+					return true
+				default:
+					errs = errors.Join(fmt.Errorf("invalid MetricType %d", m.Type()))
+					return false
+				}
+			})
+			return sm.Metrics().Len() == 0
+		})
+		return rm.ScopeMetrics().Len() == 0
+	})
+
+	if err := p.nextConsumer.ConsumeMetrics(ctx, md); err != nil {
+		errs = errors.Join(errs, err)
+	}
+
+	return errs
+}
+
+func aggregateDataPoints[DPS metrics.DataPointSlice[DP], DP metrics.DataPoint[DP]](dataPoints DPS, state *staleness.Staleness[metrics.StreamDataPoint[DP]], res pcommon.Resource, resSchemaURL string, scope pcommon.InstrumentationScope, scopeSchemaURL string, m pmetric.Metric) {
+	metric := metrics.From(res, resSchemaURL, scope, scopeSchemaURL, m)
+	metricID := metric.Identity()
+
+	now := time.Now()
+
+	for i := 0; i < dataPoints.Len(); i++ {
+		dp := dataPoints.At(i)
+
+		streamDataPointID := metrics.StreamDataPointIdentity(metricID, dp)
+
+		existing, ok := state.Load(streamDataPointID)
+		if !ok {
+			state.Store(streamDataPointID, metrics.StreamDataPoint[DP]{
+				Metric:      metric,
+				DataPoint:   dp,
+				LastUpdated: now,
+			})
+			continue
+		}
+
+		// Check if the datapoint is newer
+		if dp.Timestamp().AsTime().After(existing.DataPoint.Timestamp().AsTime()) {
+			state.Store(streamDataPointID, metrics.StreamDataPoint[DP]{
+				Metric:      metric,
+				DataPoint:   dp,
+				LastUpdated: now,
+			})
+			continue
+		}
+
+		// Otherwise, we leave existing as-is
+	}
+}
+
+func (p *Processor) exportMetricsLoop() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-p.exportTicker.C:
+			p.exportMetrics()
+		}
+	}
+}
+
+func (p *Processor) exportMetrics() {
+	p.stateLock.Lock()
+	defer p.stateLock.Unlock()
+
+	// We have to generate our own metrics slice to send to the nextConsumer
+	md := pmetric.NewMetrics()
+
+	// We want to avoid generating duplicate ResourceMetrics, ScopeMetrics, and Metrics
+	// So we use lookups to only generate what we need
+	rmLookup := map[identity.Resource]pmetric.ResourceMetrics{}
+	smLookup := map[identity.Scope]pmetric.ScopeMetrics{}
+	mLookup := map[identity.Metric]pmetric.Metric{}
+
+	// TODO: Once the upcoming RangeFunc Experiment is fully available, we can switch these to
+	// use `range` syntax. See: ttps://go.dev/wiki/RangefuncExperiment
+
+	p.numbers.Items()(func(dataID identity.Stream, dp metrics.StreamDataPoint[pmetric.NumberDataPoint]) bool {
+		m := getOrCreateMetric(dataID, dp.Metric, md, rmLookup, smLookup, mLookup)
+
+		sum := m.Sum()
+		numDP := sum.DataPoints().AppendEmpty()
+		dp.DataPoint.CopyTo(numDP)
```

Comment:
We need to update the timestamp to `now()`. Now() should be calculated once at the top and the same value used for everything

Reply:
I don't think we should set any timestamps to "now". We are aggregating data points so should use the timestamps they carry in a meaningful way. The appropriate behavior may depend on the type of metric, but for example if aggregating a sum, I would expect the start time to be the earliest from all data points being aggregated, and the end time to be the latest from those data points.

Reply:
If we don't update the timestamp, then why are we continuing to re-export them? Depending on the final destination, we're either creating duplicate data, or requiring the destination to implement data deduplication.

Reply:
It's not clear to me why we would continue to re-export data points when there is no new data to report. That said, I haven't had time to focus on the nuances of this processor and likely won't for the foreseeable future so I'll leave it to others to decide. It just seems to me that we should be reporting a direct representation of the data points we actually receive.

Reply:
Right. What you're describing is then similar to the "second" approach I describe here: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29461#issuecomment-1957467007


I think both approaches are "valid". But we just need to decide which this processor is doing

Reply:
Thanks for pointing that out. The second approach sounds better to me.

Reply:
For us, the principal use case of this set of features is to be able to send count connector metrics to a Prometheus remote write endpoint.

The count connector sends a stream of increment data points (deltas with value one).

Prometheus expects a steady stream of cumulative data points, published at the "scrape interval".

For a Prometheus scrape, consecutive reads of a counter which hasn't incremented will read the same value each time, so we see the behavior of a steady stream of repeating, unchanging values, every interval.

It is my understanding that downstream metrics systems based on Prometheus expects to see this steady stream of unchanging cumulatives in this case, and if there is a gap they will see this as a break in the series and may not behave as expected.

Reply:
Looking at the behaviour of the collector's own metrics, when the service's telemetry is configured to export  metrics via a periodic reader as follows:
```service:
  telemetry:
    metrics:
      level: detailed
      readers:
        - periodic:
            interval: 15000
            exporter:
              otlp:
                endpoint: localhost:4321
                protocol: grpc/protobuf
```
we see Sum metrics exported at every interval, even when there is no change in value. For example, I see a series of `0` values for metric `otelcol_exporter/send_failed_metric_points`.

Therefore, I think we somehow need to be able to get this behaviour in the count connector scenario described above (or any case where source metric datapoints are received only on every change in value).  

To be clear, I mean here to transform a stream of datapoints received at a variable frequency into a stream of datapoints at a fixed frequency, continuing to publish new datapoints even when there is no change in value, as we see in both the Prometheus scrape case, and the collector's own metrics case.

I guess we could rescope this to only publish datapoints on change, but if we do that, we would then need _another_ process to meet the need described here, and it's not quite clear to me how that other processor could work if it wasn't also doing what this processor would do.

I'm not clear what the concrete use case is for _not_ publishing new datapoints when there has not been a change in value, but if there _is_ a use case for both modes of operation, then maybe this could be selected by a config option?

Reply:
I think the general idea with _not_ publishing when there is no new information is that, in theory at least, it's not useful.

If we are sure that backends expect the no-new-info points, then we should probably include them in this processor's output. However, I am surprised to here this because I have not seen this in my experience (though I have barely touched prometheus) and because @sh0rez's [comment here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29461#issuecomment-1973213815) indicates that prometheus does not need these data points.

Reply:
Prometheus will consider a timeseries "stale" if it doesn't see any data points for some time (default 5 minutes). See here: https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness

> If no sample is found (by default) 5 minutes before a sampling timestamp, no value is returned for that time series at this point in time. This effectively means that time series "disappear" from graphs at times where their latest collected sample is older than 5 minutes or after they are marked stale.

Imagine you have an alert configured in Prometheus Alert Manager based on the ratio of the collector's metrics `otelcol_exporter_send_failed_metric_points` and `otelcol_exporter_sent_metric_points`, say to trigger an alert when the ratio of failed is greater than 10 percent.

Prometheus can only calculate the ratio where it has values for both series, so following the staleness rule above, only up to five minutes since the last datapoint recorded for **both** series.

We can see how we could easily get in trouble if we didn't publish new data points for a series that didn't change in this case.

If `otelcol_exporter_sent_metric_points` stopped incrementing, and `otelcol_exporter_send_failed_metric_points` _did_ start to increment, then depending on the time window used to calculate the ratio etc, it could take more than five minutes for the ratio to exceed 10 percent.  So, if we don't continue to have new datapoints for the unchanging `otelcol_exporter_sent_metric_points` series, we wouldn't actually trigger the alert in this case.

Of course this is a synthetic example, but I think it illustrates how the convention of Prometheus datasources continuing to periodically produce datapoints even when a series' value  hasn't changed, is sometimes relied upon.